Src/ReadBarometer.c

```diff
@@ -6,127 +17,244 @@
 
+        // Read the second byte (bits 15-8)
         HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
         temperatureReading += dataInBuffer << 8;
 
+        // Read the third byte (bits 7-0)
         HAL_SPI_TransmitReceive(&hspi2, &ADC_READ_CMD, &dataInBuffer, CMD_SIZE, CMD_TIMEOUT);
         temperatureReading += dataInBuffer;
 
+        // Release chip select to end the command
         HAL_GPIO_WritePin(BARO_CS_GPIO_Port, BARO_CS_Pin, GPIO_PIN_SET);
 
-        // calculate 1st order pressure and temperature (1st order algorithm from MS5607-02BA03 datasheet)
-        double dT = temperatureReading - c5Tref * pow(2, 8);
-        double temp = 2000 + dT * c6Tempsens / pow(2, 23);
-        double off = c2Off * pow(2, 17) + dT * c4Tco / pow(2, 6);
-        double sens = c1Sens * pow(2, 16) + dT * c3Tcs / pow(2, 7);
-        // pressure and temperature were not divided by 100 to keep the decimal places
-        double pressure = (pressureReading * sens / pow(2, 21) - off) / pow(2, 15); // need to divide P by 100 to get mbar
-        double temperature = (temp); // need to divide T by 100 to get degrees Celcius
+        /* Calculate First-Order Temperature and Pressure --------------------*/
+
+        // Equations provided by MS5607-02BA03 data sheet
+
+        int32_t dT      = temperatureReading - (c5Tref << 8);
+        int32_t temp    = 2000 + ((dT * c6Tempsens) >> 23);                 // Divide this value by 100 to get degrees Celcius
+        int64_t off     = (c2Off << 17) + ((dT * c4Tco) >> 6);
+        int64_t sens    = (c1Sens << 16) + ((dT * c3Tcs) >> 7);
+        int32_t p       = (((pressureReading * sens) >> 21) - off) >> 15;   // Divide this value by 100 to get millibars
+
+        /* Calculate Second-Order Temperature and Pressure -------------------*/
+
+        // Equations provided by MS5607-02BA03 data sheet
+
+        if (temp < TEMP_LOW)    // If the temperature is below 20°C
+        {
+            int32_t t2      = (dT * dT) >> 31;
+            int32_t off2    = 61 * (((temp - 2000) * (temp - 2000)) >> 4);
+            int32_t sens2   = 2 * ((temp - 2000) * (temp - 2000));
+
+            if (temp < TEMP_VERY_LOW)   // If the temperature is below -15°C
+            {
+                off2    = off2 + (15 * ((temp + 1500) * (temp + 1500)));
+                sens2   = sens2 + (8 * ((temp + 1500) * (temp + 1500)));
+            }
+
+            temp    = temp  - t2;
+            off     = off   - off2;
+            sens    = sens  - sens2;
+        }
+
+        /* Store Data --------------------------------------------------------*/
 
-        if (osMutexWait(data->mutex_, 0) != osOK)
+        if (osMutexWait(data->mutex_, 0) == osOK)
         {
-            continue;
+            // Assign new values if mutex was available
+            data->pressure_     = p;
+            data->temperature_  = temp;
+            osMutexRelease(data->mutex_);
```

Comment:
I am just wondering what scenario could lead this if statement not to be called?

Reply:
Mmm, for an example, consider we decide that if the pressure spikes, we want to ignore it, someone might write.

```C
        if (osMutexWait(data->mutex_, 0) == osOK && pressure < 1000)
        {
            // Assign new values if mutex was available
            data->pressure_     = p;
            data->temperature_  = temp;
            osMutexRelease(data->mutex_);
        }
```

Reply:
Actualy y'know, taking a second look at this, I think I made a comment on this more because I was more a fan of how it was written before? Less nested stuff is easier to read, and errors should be treated like special cases, I like to keep what "usually" happens in the regular flow of code. All in all, I guess commenting on this is pretty trivial anyways so it's not really important to change this.

Reply:
I know that using `continue` can be confusing to newer programmers, and it's apparently bad practice (although I am a quite a big fan and use of it regularly... oops). But this just stood out to me as something that just flowed better when reading throught the code. It also removes an operation.

As for if the pressure spikes, I would then write the code as such:
```c
if (pressure < 1000 && osMutexWait(data->mutex_, 0) == osOK)
{
    data->pressure_     = p;
    data->temperature_  = temp;
    osMutexRelease(data->mutex_);
}
````
Putting the `pressure < 1000` condition first means that (at least in Java, I'm not 100% sure in C) if `pressure < 1000` fails, then `osMutexWait(data->mutex_, 0) == osOK` is never called, as the whole condition will be false anyways, meaning the mutex is never signed out. If C doesn't work like this, then I would just nest the entire if block into another where its condition is solely `if (pressure < 1000)`.

Reply:
We don't actually want to do that lol, it was just an example regarding expandability moving forward.